detekt-rules-coroutines/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/coroutines/SuspendFunSwallowedCancellation.kt

```diff
@@ -0,0 +1,180 @@
+ *         bar(1_000L)
+ *     }
+ * }
+ * </noncompliant>
+ *
+ * <compliant>
+ * @@Throws(IllegalStateException::class)
+ * suspend fun bar(delay: Long) {
+ *     check(delay <= 1_000L)
+ *     delay(delay)
+ * }
+ *
+ * suspend fun foo() {
+ *     try {
+ *         bar(1_000L)
+ *     } catch (e: IllegalStateException) {
+ *         // handle error
+ *     }
+ * }
+ *
+ * // Alternate
+ * @@Throws(IllegalStateException::class)
+ * suspend fun foo() {
+ *     bar(1_000L)
+ * }
+ * </compliant>
+ *
+ */
+@RequiresTypeResolution
+class SuspendFunSwallowedCancellation(config: Config) : Rule(config) {
+    override val issue = Issue(
+        id = "SuspendFunInsideRunCatching",
+        severity = Severity.Minor,
+        description = "`suspend` functions should not be called inside `runCatching`'s lambda block, because " +
+            "`runCatching` catches all the `Exception`s.",
+        debt = Debt.TEN_MINS
+    )
+
+    override fun visitCallExpression(expression: KtCallExpression) {
+        super.visitCallExpression(expression)
+
+        val resultingDescriptor = expression.getResolvedCall(bindingContext)?.resultingDescriptor
+        resultingDescriptor ?: return
+        if (resultingDescriptor.fqNameSafe != RUN_CATCHING_FQ) return
+
+        expression.forEachDescendantOfType<KtCallExpression>({
+            expression == it || shouldTraverseInside(it, bindingContext)
+        }) { descendant ->
+            val callableDescriptor = descendant.getResolvedCall(bindingContext)?.resultingDescriptor
+            if (callableDescriptor?.isSuspend == true) {
+                report(
+                    message = "The suspend function call ${callableDescriptor.fqNameSafe.shortName()} called inside " +
+                        "`runCatching`. You should either use specific `try-catch` only catching exception that you " +
+                        "are expecting or rethrow the `CancellationException` if already caught.",
+                    expression
+                )
+            }
+        }
+
+        expression.forEachDescendantOfType<KtForExpression> { descendant ->
+            if (descendant.doesLoopHasSuspendingIterators()) {
+                report(
+                    message = "The for-loop expression has suspending operator which is called inside " +
+                        "`runCatching`. You should either use specific `try-catch` only catching exception that you " +
+                        "are expecting or rethrow the `CancellationException` if already caught.",
+                    expression
+                )
+            }
+        }
+    }
+
+    @Suppress("ReturnCount")
+    private fun shouldTraverseInside(psiElement: PsiElement, bindingContext: BindingContext): Boolean {
+        if (psiElement is KtCallExpression) {
+            val callableDescriptor =
+                psiElement.getResolvedCall(bindingContext)?.resultingDescriptor as? FunctionDescriptor
+            callableDescriptor ?: return false
```

Comment:
this can be merged together to prevent strange "return on variable":
```suggestion
            val callableDescriptor =
                (psiElement.getResolvedCall(bindingContext)?.resultingDescriptor as? FunctionDescriptor)
                ?: return false
```
alternatively match the `if-return` style below instead of elvis.